[DOC] Configure broker certificates per listener (#2414)
strimzi/strimzi-kafka-operator: documentation/modules/security/ref-formats-kafka-listener-certificates.adoc

// Module included in the following assemblies:
//
// assembly-security.adoc

[id='ref-formats-kafka-listener-certificates-{context}']

> = SAN certificates for Kafka listeners

Reviewer: ```suggestion
= Alternative subjects in server certificates for Kafka listeners
```

> If you provide your own xref:kafka-listener-certificates-{context}[Kafka listener certificates], you must use a Subject Alternative Name (SAN) certificate for each listener. The SAN certificate must specify hostnames for:

Reviewer: Sorry if I confused you about this. The SANs are a feature/configuration  of the certificate. It is not certificate type. So I think it should sound something like this:

```suggestion
In order to use TLS hostname verification with your own xref:kafka-listener-certificates-{context}[Kafka listener certificates], you must use the correct Subject Alternative Names (SAN) for each listener. The certificate SANs must specify hostnames for:
```

* All of the Kafka brokers in your cluster

* The Kafka cluster bootstrap service

> The Kafka broker hostnames can contain wildcard names.

Reviewer: ```suggestion
You can use wildcsard certificates if supported by your CA.
```

> Using SAN certificates is required for TLS hostname verification.

Reviewer: ```suggestion
Using certificates with the coorrect SANs is required for TLS hostname verification.
```

Author: I'll remove this sentence because we now mention TLS hostname verification right at the start.

> == TLS listener SAN certificate examples

Reviewer: ```suggestion
== TLS listener SAN examples
```

> Use the following examples to help you specify hostnames in SAN certificates for TLS listeners.

Reviewer: ```suggestion
Use the following examples to help you specify hostnames the SANs in your certificates for TLS listeners.
```

.Wildcards example

[source,shell,subs="+quotes,attributes+"]
----
//Kafka brokers
*._<cluster-name>_-kafka-brokers
*._<cluster-name>_-kafka-brokers._<namespace>_.svc

// Bootstrap service
_<cluster-name>_-kafka-bootstrap
_<cluster-name>_-kafka-bootstrap._<namespace>_.svc
----

.Non-wildcards example

[source,shell,subs="+quotes,attributes+"]
----
// Kafka brokers
_<cluster-name>_-kafka-0._<cluster-name>_-kafka-brokers
_<cluster-name>_-kafka-0._<cluster-name>_-kafka-brokers._<namespace>_.svc
_<cluster-name>_-kafka-1._<cluster-name>_-kafka-brokers
_<cluster-name>_-kafka-1._<cluster-name>_-kafka-brokers._<namespace>_.svc
# ...

// Bootstrap service
_<cluster-name>_-kafka-bootstrap
_<cluster-name>_-kafka-bootstrap._<namespace>_.svc
----

> == External listener SAN certificate examples

Reviewer: ```suggestion
== External listener SAN examples
```

> For external listeners which have TLS encryption enabled, the hostnames you need to specify in SAN certificates depends on the external listener `type`.

Reviewer: ```suggestion
For external listeners which have TLS encryption enabled, the hostnames you need to specify in certificates depends on the external listener `type`.
```

[cols="2*",options="header",stripes="none",separator=¦]
|===

¦External listener type

> ¦In the SAN certificate, specify...

Reviewer: ```suggestion
¦In the SANs, specify...
```

m¦Route

> ¦Addresses of all Kafka broker `Routes` and the address of the bootstrap service `Route`.

Reviewer: ```suggestion
¦Addresses of all Kafka broker `Routes` and the address of the bootstrap `Route`.
```

You can use a matching wildcard name.

m¦loadbalancer

> ¦Addresses of all Kafka broker `loadbalancers` and the `loadbalancer` bootstrap address.

Reviewer: ```suggestion
¦Addresses of all Kafka broker `loadbalancers` and the bootstrap `loadbalancer` address.
```
